### DISCUSSION TITLE
Can you train on non-square images?

### DISCUSSION THREAD

**Discussion by maximxlss** [d8ahazard/sd_dreambooth_extension#943] [Q&A]
Recently, I've been trying to get training working. I've trained with this ext successfully in the past, but this time nothing works (outputs stay the same, can't remember how many steps was it, but half an hour on 3090 ti, I think about a thousand). The thing is, this time I haven't cropped the images (and I would prefer not to). Textual inversion and hypernetworks don't work too (outputs garbage), so I suspect that the aspect ratio is to blame. Could this be the case, and if so, what's the best way to solve this?

**Comment by Arilziem** [comment]
Uncropped works fine. There have been issues with xformers 0.16, so make sure you are not using that (use the older 0.14 or the newer 0.17 instead). Maybe also share your training configuration and the version info you get when starting dreambooth?

**Reply by 78Alpha** [reply]
All images get cropped when training starts

`for file in concept_images:
                try:
                    img = Image.open(file)
                    if img.width != size or img.height != size:
                        needs_crop = True
                except Exception as e:
                    print(f"Exception parsing instance image: {e}")
                    continue`

For odd shapes, you have to make squares but with padding. Otherwise you get the issues the base model has.

**Reply by Arilziem** [reply]
No, that is false. Images are loaded using bucketing. Have a look at [finetune_utils.generate_dataset](https://github.com/d8ahazard/sd_dreambooth_extension/blob/21ddebabb47274255d516e655f997d9ddadc0fe4/dreambooth/finetune_utils.py#L855) / [DbDataset](https://github.com/d8ahazard/sd_dreambooth_extension/blob/21ddebabb47274255d516e655f997d9ddadc0fe4/dreambooth/finetuning_dataset.py#L19) instead.

The only thing that still uses SuperDataset in main branch is when you click the "Preview Prompts" button as far as I can tell. It has been removed in dev branch.

**Reply by 78Alpha** [reply]
That is certainly useful to see implemented. It seems to drop some images that can't be scaled to what it wants. If your set is small and it drops them all, may be an issues, but there are cropping and padding tools for that case.

**Reply by Arilziem** [reply]
It is indeed very useful. :D What makes you say that it drops images?

**Reply by 78Alpha** [reply]
These lines from the D8ahazaard version

```
        assert image.shape[0] == reso[1] and image.shape[1] == reso[0], \
            f"internal error, illegal trimmed size: {image.shape}, {reso}"
        return image
```

and this snippet from the original NovelAI [ReadME](https://github.com/NovelAI/novelai-aspect-ratio-bucketing)

> gen_buckets: 0.00012s
> skipped images: 344
> aspect error: mean 0.03291049121627481, median 0.022727272727272707, max 3.991769547325103

**Reply by Arilziem** [reply]
I would encourage you to actually try and verify before suggesting things are broken. That assert comes after trimming so I'm not sure why it's even there, the result should always be correct. I guess for debugging. In any case, dev has finally been merged into main with a lot of improvements, the [new load and trim code](https://github.com/d8ahazard/sd_dreambooth_extension/blob/d4e2f7d5dc5c67cc6292c8b8137b688a43247c35/dreambooth/utils/image_utils.py#L418) looks a bit different now.

**Comment by d8ahazard** [comment]
Yeah, there is definitely no dropping of images going on. The bit of assert code in the trimming logic is just there to ensure everything is trimmed as intended. All of your instance images should be used regardless of scale. If you train using perfectly square images - then no cropping or trimming will occur. If you use non-square images, they will be scaled as a factor of the maximum resolution, with buckets created for the various "common" aspect ratios around that. 

Any cropping that happens should be a minimum, as I perform a combination of upscaling and cropping to try and fit the images to the target bucket(s).
